### ISSUE TITLE
[Feature Request] Support for DW2PDF

### ISSUE THREAD

**Issue by vladipus** [restlessminded/dokuwiki-plugin-svgEmbed#9]
Is there any chance for this great PDF export plugin to be supported?
https://www.dokuwiki.org/plugin:dw2pdf

Thank you for your work.

**Comment by restlessminded** [comment]
I see they have documentation on supporting their plugin.  I'll have to read it over and see what's involved and what I can do, but I agree that supporting PDF export would be pretty useful.

**Comment by vladipus** [comment]
> I see they have documentation on supporting their plugin. I'll have to read it over and see what's involved and what I can do, but I agree that supporting PDF export would be pretty useful.

Even more to this, is that their backend mPDF actually supports SVGs! As far as I could think, the implementation should be pretty trivial...

**Comment by vladipus** [comment]
I've just checked it. If I embed the SVG with an <img> tag, everything works fine and the output PDF (generated by mpdf) contains the SVG image in question.

It is possible to override the default behavior of <object><svg> and use img instead?

**Comment by restlessminded** [comment]
Checking in:  I'm still going to look into this.  I just haven't had a good opportunity for some extended time to look at it.  Working form home means that I sit here for nine or ten hours a day doing stuff for work and I give the computer hairy eyeball most of the rest of the time.  👁️

**Comment by vladipus** [comment]
I feel ya, man. Take your time. But all that is really needed from you is to provide some kind of tag option to go the easy way - and <img> instead of <object> embedding. That should be easier of what you've already done.

I'm ready to test this for you, just in case.

**Comment by restlessminded** [comment]
Since I had another issue come in I figured, "I'm in here, I might as well knock this out."  I think it should be working, please update and test.

I am a bit embarrassed to say that it wasn't working and I couldn't figure out what the problem was.  In the end it was that I hadn't touched the configuration files to invalidate Dokuwiki's cache so it was using old rendered data.  You might want to do the same after updating just in case.

This might not be the final product.  I want to know how your test goes first.

**Comment by vladipus** [comment]
Great job! Works out of the box just after the update.
A kinda strange issue is present however, which doesn't output some of the vectors.
https://turbanov.ru/wiki/en/toolworks/docs/apparatus/steady-tick
You can export this page to a PDF to check this out.
Might be a DW2PDF issue actually. Will experiment more on that.

**Comment by vladipus** [comment]
Turns out that even Acrobat doesn't read this correctly so it's might be of some unsupported features. Don't use gradients I would say. Will have to redesign some stuff, oh well.

Thanks again.

**Comment by restlessminded** [comment]
Don't

**Comment by restlessminded** [comment]
Okay, I'm back at the computer and I see what you mean about gradients not being rendered.  Let me work on that.

I also see that SVGs are being rendered at the full width.  I figured that was going to happen, so I need to investigate that, too.

**Comment by vladipus** [comment]
Roger that.
I just suppose that modern day PDF renderers have some limitations. Maybe I'm wrong. Maybe we should try converting a simple page with mpdf directly? To see if it's a weak spot.

Acrobat on mobiles is rendering even worse...
I think the problem is some kind of limit for using gradients to fill the objects.

**Comment by restlessminded** [comment]
Right now I am considering offering an option to rasterize the file.  That would require the imagick PHP plugin, though.

First I am trying to fix the output sizing the SVG.  For display I built a `<span>` that contains the SVG and that works fine for printing from the browser, too, but export isn't working quite so well.

**Comment by restlessminded** [comment]
Okay, I have managed to figure out:

- mPDF uses GD to do image conversions, which won't keep your alpha transparency
- ImageMagick _will_ keep the gradients, but only if you install Inkscape as well (`convert` ends up calling Inkscape to do the actual work)
- I'm still tinkering with spacing

**Comment by restlessminded** [comment]
Actually, rsvg-convert maintains the alpha channel on conversion, too.  Now I just have to look into adding an image to the cache, but I have some other things to attend to.  I think I we can make this work out for you.

**Comment by restlessminded** [comment]
Actually, [it's documented that alpha is not supported in any image format.](https://www.dokuwiki.org/plugin:dw2pdf#linked_images)  I guess that means we're out of luck since PNG alpha isn't supported either.

I am still going to work on the spacing, though.  Making all SVG files 100% width is space-wasteful and not true to the source.

**Comment by vladipus** [comment]
Not exactly. It's not about alpha, really as the main issue seems to be related to gradients used. You can actually see semi-transparent circles and the work just fine. Only gradients fail (some of them).

**Comment by restlessminded** [comment]
Actually, I've been bug-fixing and was revisiting this; I don't know about installing Inkscape to work with Imagick, especially since I suspect a large proportion of servers won't have a GUI and it wants to install a quarter gig of X11 to move forward with that.

Then again, I haven't tested it (maybe it renders it in-app, Imagick has had tons of updates since 2021).  Hmm...

**Comment by restlessminded** [comment]
I just tried `convert` and it strips out the alpha, as per normal.  I think in this case we might have to call it best-effort.

**Comment by restlessminded** [comment]
Okay, we can do this:  if you put a .png that you generate of your .svg file (with, say, `rsvg-convert`) with the same base name in the same namespace then when you use dw2pdf it will choose to pull the .png file instead of the .svg file.  That way you can make the images with gradients as a .png file, which is supported by dw2pdf.  I am committing the change now.
